ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -0,0 +1,489 @@
+    rb_raise(rb_eRuntimeError, "Duplicate heap object tracking: %lu", obj);
+    return;
+  }
+
+  fprintf(stderr, "Committed allocation of %lu (heap_record=%p, object_record=%p)\n", obj, heap_record, object_record);
+
+  heap_record->num_tracked_objects++;
+  heap_record->inuse_objects += weight;
+}
+
+void commit_free(heap_recorder *heap_recorder, VALUE obj) {
+  st_data_t key = (st_data_t) obj;
+  object_record *object_record = NULL;
+  if (!st_delete(heap_recorder->object_records, (st_data_t*) &key, (st_data_t*) &object_record)) {
+    // Object not tracked?
+    rb_raise(rb_eRuntimeError, "Committing free of untracked object");
+    return;
+  }
+
+  heap_record *heap_record = object_record->heap_record;
+  heap_record->num_tracked_objects--;
+  heap_record->inuse_objects -= object_record->weight;
+
+  fprintf(stderr, "Committed free of %lu (heap_record=%p, object_record=%p)\n", obj, heap_record, object_record);
+
+  object_record_free(object_record);
+}
+
+static void flush_queue(heap_recorder *heap_recorder) {
+  for (size_t i = 0; i < heap_recorder->queued_samples_len; i++) {
+    sample *queued_sample = &heap_recorder->queued_samples[i];
+    if (!queued_sample->skip) {
+      fprintf(stderr, "Flushing sample %p\n", queued_sample);
+
+      if (queued_sample->free) {
+        commit_free(heap_recorder, queued_sample->obj);
+      } else {
+        commit_allocation(heap_recorder, queued_sample->stack, queued_sample->obj, queued_sample->weight);
+      }
+    }
+
+    *queued_sample = EmptySample;
+  }
+  heap_recorder->queued_samples_len = 0;
+}
+
+static void enqueue_sample(heap_recorder *heap_recorder, sample new_sample) {
+  fprintf(stderr, "Enqueuing sample for %lu (weight=%u free=%i)\n", new_sample.obj, new_sample.weight, new_sample.free);
+  if (heap_recorder->queued_samples_len >= MAX_QUEUE_LIMIT) {
+    fprintf(stderr, "Dropping sample on the floor.\n");
+    return;
+  }
+
+  heap_recorder->queued_samples[heap_recorder->queued_samples_len] = new_sample;
+  heap_recorder->queued_samples_len++;
+}
+
+static void enqueue_allocation(heap_recorder *heap_recorder, heap_stack *heap_stack, VALUE obj, unsigned int weight) {
+  enqueue_sample(heap_recorder, (sample) {
+      .stack = heap_stack,
+      .obj = obj,
+      .weight = weight,
+      .free = false,
+      .skip = false,
+  });
+}
+
+static void enqueue_free(heap_recorder *heap_recorder, VALUE obj) {
+  enqueue_sample(heap_recorder, (sample) {
+      .stack = NULL,
+      .obj = obj,
+      .weight = 0,
+      .free = true,
+      .skip = false,
+  });
+}
+
+void start_heap_allocation_recording(heap_recorder* heap_recorder, VALUE new_obj, unsigned int weight, ddog_CharSlice *class_name) {
+  fprintf(stderr, "Started recording allocation of %lu with weight %u\n", new_obj, weight);
+  partial_heap_recording *active_recording = &heap_recorder->active_recording;
+  active_recording->obj = new_obj;
+  active_recording->weight = weight;
+  active_recording->class_name = class_name;
```

Comment:
Minor: I usually favor this style to fill up structures, since it has slightly less typing and has the advantage that anything you don't set gets set to zero, rather than being accidental garbage (insert snarky remark about being surprised in C about getting sane behavior) 

```suggestion
  heap_recorder->active_recording = (partial_heap_recording) {
    .obj = new_obj,
    .weight = weight,
    .class_name = class_name,
  }
```